[6.0 🍒] fix reverse cond-fail for SuppressedAssociatedTypes

kavon opened the pull request:
- Explanation: Allows the compiler to ingest older stdlib interface files that use ~Copyable on associated types, without those stdlibs being built with the experimental feature that it's now gated-behind.
- Scope: Changes how we infer a feature flag.
- Issue: rdar://126668752
- Original PR: https://github.com/apple/swift/pull/73111
- Risk: Virtually none.
- Testing: Building the stdlib should be sufficient.
- Reviewer: @tshortli

tshortli commented on stdlib/cmake/modules/SwiftSource.cmake:
```diff
@@ -622,7 +622,6 @@ function(_compile_swift_files
   endif()
 
   list(APPEND swift_flags "-enable-experimental-feature" "NoncopyableGenerics2")
-  list(APPEND swift_flags "-enable-experimental-feature" "SuppressedAssociatedTypes")
```
I think it would be clearer to leave this in, personally. Also, usually we solve this class of problems by relaxing the diagnostics if the source file kind is a swiftinterface, rather than by enabling the feature.

kavon committed:
NCGenerics: fix reverse condfail

I gated inverses on associated types behind a feature flag, but a
slightly older stdlib being rebuilt from source will now fail because it
 used that functionality for `_Pointer` without the feature.

So, just permit the feature if we're working with a module built from
an interface file, to avoid this reverse condfail
.

(cherry picked from commit 779184f9220e9bbcd31435ead7d373654ba1ed0e)

kavon commented:
@swift-ci test

kavon commented:
@swift-ci test
